FTPFolderUploader/main.py

```diff
@@ -0,0 +1,293 @@
+#!/usr/bin/env python
+#  -*- coding: utf8 -*-
+
+# derived from https://stackoverflow.com/questions/5663787/upload-folders-from-local-system-to-ftp-using-python-script
+
+
+import configparser
+import os
+import getpass
+import sys
+import time
+import paramiko
+
+SLEEP_SECONDS = 10
+REMOTE_INCOMING_DIR = "staging"
+
+
+def _get_local_files(local_dir, walkdir=False):
+
+    '''Retrieve local files list
+        result_list == a list of dictionaries with path and mtime keys. ex: {'path':<filepath>,'mtime':<file last modified time>}
+        ignore_dirs == a list of directories to ignore, should not include the base_dir.
+        ignore_files == a list of files to ignore.
+        ignore_file_ext == a list of extentions to ignore.
+    '''
+
+    result_list = []
+    ignore_dirs = ['CVS', '.svn', '.git']
+    ignore_files = ['.project', '.pydevproject', '.gitignore', '.DS_Store']
+    ignore_file_ext = ['.pyc']
+
+    base_dir = os.path.abspath(local_dir)
+
+    for current_dir, dirs, files in os.walk(base_dir):
+        for this_dir in ignore_dirs:
+            if this_dir in dirs:
+                dirs.remove(this_dir)
+
+        sub_dir = current_dir.replace(base_dir, '')
+        if not walkdir and sub_dir:
+            break
+
+        for this_file in files:
+            if this_file not in ignore_files and os.path.splitext(this_file)[-1].lower() not in ignore_file_ext:
+                filepath = os.path.join(current_dir, this_file)
+                file_monitor_dict = {
+                                     'path': filepath,
+                                     'mtime': os.path.getmtime(filepath)
+                                     }
+                result_list.append(file_monitor_dict)
+    return result_list
+
+
+def monitor_and_ftp(server,
+                      username,
+                      password,
+                      local_dir,
+                    walkdir=False):
+
+    last_files_list = _get_local_files(local_dir, walkdir)
+
+    '''Monitor local files and when an update is found connect and upload'''
+    print('\nMonitoring changes in (%s).' % (os.path.abspath(local_dir)))
+    print('(Use ctrl-c to exit)')
+
+    while True:
+        try:
+
+            latest_files_list = _get_local_files(local_dir, walkdir)
+
+            files_to_update = []
+
+            for idx in range(len(latest_files_list)):
+
+                if idx < len(last_files_list):
+                    # compare last modified times
+                    if latest_files_list[idx]['mtime'] > last_files_list[idx]['mtime']:
```

Comment:
Doing this comparison by index may not be reliable, since os.walk does not guarantee ordered results (since it calls os.scandir for results, see below), so the same index may not be the same file. You really need to compare files by name.

https://docs.python.org/3/library/os.html#os.scandir

Reply:
trap for young players